### DISCUSSION TITLE
QrCode Dinâmico - Erro de leituras

### DISCUSSION THREAD

**Discussion by minoscorpion** [bacen/pix-api#609] [Q&A]
Boa tarde ! . 

Estamos enfrentando alguns problemas com a leitura de nosso qrcode em produção por parte de alguns bancos (Nubank, c6, etc), outros conseguimos ler normalmente (itau, bradesco, caixa, santander)

Porem esta difícil a analise e como descobrir o porquê algumas instituições simplesmente não conseguem ler, algumas que conseguimos contato, simplesmente dizem que são por conta de nosso certificado, dizendo haver um erro no tls (efetuamos a troca do certificado no STA há uns 20 dias atras).

Pelo que olhamos intermanamente, nosso certificado no STA foi aceito sem problemas e nosso certificado para o download do base64 está valido.

Se alguém puder contribuir com ideias do que possamos analisar e/ou fazer.

Segue um emv gerado por nós ! 

00020101021226850014BR.GOV.BCB.PIX2563qr.ourinvest.com.br/pix/v2/cob/d037112f9bbf4f1dae98493f6650cef15204000053039865802BR5909Ourinvest6009Sao Paulo62070503***630468C6

**Comment by rodri-oliveira-dev** [comment]
Pelo cenário, eu separaria a análise do QR em duas partes: o payload EMV e o acesso HTTPS à URL dinâmica.

No EMV que você enviou, inclusive, o CRC fecha em `68C6`, então eu não começaria procurando problema na montagem básica do payload.

O fato de alguns bancos conseguirem consumir normalmente e outros falharem também me faria olhar primeiro para interoperabilidade TLS no endpoint:

`qr.ourinvest.com.br`

Eu validaria principalmente:

- cadeia completa do certificado enviada pelo servidor, incluindo intermediárias;
- SAN contendo exatamente o hostname;
- suporte a TLS 1.2;
- cipher suites disponíveis;
- SNI;
- validade e ordem da cadeia;
- OCSP/CRL;
- comportamento do endpoint a partir de uma máquina externa à infraestrutura.

Uma coisa importante é não confundir o certificado utilizado na integração da API Pix/STA com o certificado HTTPS apresentado pela URL que está dentro do QR Code. O fato de um certificado ter sido aceito no STA não garante que o endpoint público do QR esteja entregando uma cadeia TLS compatível com todos os clientes.

Eu começaria com algo como:

openssl s_client -connect qr.ourinvest.com.br:443 -servername qr.ourinvest.com.br -showcerts

e conferiria se o servidor realmente entrega toda a cadeia.

Já vi esse tipo de comportamento em integrações onde navegador e alguns clientes funcionam porque conseguem completar uma intermediate certificate ausente, enquanto clientes mais restritivos simplesmente rejeitam o handshake.

Se Nubank/C6 informaram especificamente erro de TLS, eu tentaria obter deles o erro exato do handshake. Algo como `unknown_ca`, `certificate_unknown`, `handshake_failure` ou incompatibilidade de protocolo reduz bastante o espaço de investigação.
